Isolated observer copy of stella_vslam e445b545 PnP; namespace renamed to avoid interposition.

--- a/src/stella_vslam/solve/pnp_solver.h
+++ b/src/stella_vslam/solve/pnp_solver.h
@@ -1,5 +1,5 @@
-#ifndef STELLA_VSLAM_SOLVE_PNP_SOLVER_H
-#define STELLA_VSLAM_SOLVE_PNP_SOLVER_H
+#ifndef SV_OBSERVED_PNP_SOLVER_H
+#define SV_OBSERVED_PNP_SOLVER_H
 
 #include "stella_vslam/util/converter.h"
 #include "stella_vslam/type.h"
@@ -8,7 +8,7 @@
 #include <random>
 
 namespace stella_vslam {
-namespace solve {
+namespace reloc_reference {
 
 class pnp_solver {
 public:
@@ -144,4 +144,4 @@
 } // namespace solve
 } // namespace stella_vslam
 
-#endif // STELLA_VSLAM_SOLVE_PNP_SOLVER_H
+#endif // SV_OBSERVED_PNP_SOLVER_H
--- a/src/stella_vslam/solve/pnp_solver.cc
+++ b/src/stella_vslam/solve/pnp_solver.cc
@@ -1,4 +1,5 @@
-#include "stella_vslam/solve/pnp_solver.h"
+#include "pnp_solver.h"
+#include "pnp_trace.hpp"
 #include "stella_vslam/util/fancy_index.h"
 #include "stella_vslam/util/random_array.h"
 #include "stella_vslam/util/trigonometric.h"
@@ -6,7 +7,7 @@
 #include <spdlog/spdlog.h>
 
 namespace stella_vslam {
-namespace solve {
+namespace reloc_reference {
 
 pnp_solver::pnp_solver(const eigen_alloc_vector<Vec3_t>& valid_bearings,
                        const std::vector<int>& octaves,
@@ -72,6 +73,7 @@
         // 2-1. Create a minimum set
         const auto random_indices = util::create_random_array(min_set_size, 0U, num_matches_ - 1, random_engine_);
         assert(random_indices.size() == min_set_size);
+        for(auto idx : random_indices) reloc_trace::scalar("sample", idx);
 
         min_set_bearings.clear();
         min_set_pos_ws.clear();
@@ -90,6 +92,9 @@
         // 2-3. Check inliers and compute a score
         double cost = 0.0;
         const auto num_inliers = check_inliers(rot_cw_in_sac, trans_cw_in_sac, is_inlier_match_in_sac, cost);
+        reloc_trace::scalar("cost", cost);
+        reloc_trace::scalar("inliers", num_inliers);
+        for(bool inlier : is_inlier_match_in_sac) reloc_trace::scalar("mask", inlier);
 
         // 2-4. Update the best model
         if (num_inliers > min_num_inliers_ && min_cost > cost) {
@@ -101,6 +106,7 @@
     }
 
     solution_is_valid_ = min_cost < std::numeric_limits<double>::max();
+    reloc_trace::scalar("valid", solution_is_valid_);
 
     if (!recompute || !solution_is_valid_) {
         return;
@@ -159,20 +165,27 @@
     // (Lepetit et al. in IJCV 2009)
 
     const eigen_alloc_vector<Vec3_t> control_points = choose_control_points(pos_ws);
+    reloc_trace::vecs("controls", control_points);
     const eigen_alloc_vector<Vec4_t> alphas = compute_barycentric_coordinates(control_points, pos_ws);
+    reloc_trace::vecs("alphas", alphas);
 
     // Construct M matrix according to Eq.(5) and (6)
     const MatX_t M = compute_M(bearing_vectors, alphas);
+    reloc_trace::mat("M", M);
 
     // Compute singular vectors U
     const MatRC_t<12, 12> MtM = M.transpose() * M;
+    reloc_trace::mat("MtM", MtM);
     Eigen::JacobiSVD<MatRC_t<12, 12>> SVD(MtM, Eigen::ComputeFullV | Eigen::ComputeFullU);
 
     const MatRC_t<12, 12> U = SVD.matrixU();
+    reloc_trace::mat("U12", U);
 
     // Compute L matrix and rho vector in Eq.(13)
     const MatRC_t<6, 10> L_6x10 = compute_L_6x10(U);
+    reloc_trace::mat("L", L_6x10);
     const MatRC_t<6, 1> Rho = compute_rho(control_points);
+    reloc_trace::mat("rho", Rho);
 
     Mat33_t rot_cand;
     Vec3_t trans_cand;
@@ -183,17 +196,24 @@
 
         // Solve Eq.(13)
         const Vec4_t betas = find_initial_betas(L_6x10, Rho, N);
+        reloc_trace::mat("betas", betas);
         // Minimize Eq.(15)
         const Vec4_t refined_betas = gauss_newton(L_6x10, Rho, betas, num_iter);
+        reloc_trace::mat("refined", refined_betas);
 
         // Eq.(16)
         const eigen_alloc_vector<Vec3_t> ccs = compute_ccs(refined_betas, U);
+        reloc_trace::vecs("ccs", ccs);
         const bool bearing_z_sign = bearing_vectors.at(0)(2) > 0;
         const eigen_alloc_vector<Vec3_t> pcs = compute_pcs(alphas, ccs, bearing_z_sign);
+        reloc_trace::vecs("pcs", pcs);
 
         estimate_R_and_t(pos_ws, pcs, rot_cand, trans_cand);
+        reloc_trace::mat("rotation", rot_cand);
+        reloc_trace::mat("translation", trans_cand);
 
         const auto reproj_error = reprojection_error(pos_ws, bearing_vectors, rot_cand, trans_cand);
+        reloc_trace::scalar("error", reproj_error);
 
         // Take the rotation and translation which archieve the minimum reprojection error
         if (reproj_error < reproj_minimum) {
@@ -226,9 +246,13 @@
     }
 
     const MatX_t PW0tPW0 = PW0.transpose() * PW0;
+    reloc_trace::mat("PW0", PW0);
+    reloc_trace::mat("PW0tPW0", PW0tPW0);
     Eigen::JacobiSVD<MatX_t> SVD(PW0tPW0, Eigen::ComputeFullV | Eigen::ComputeFullU);
     const MatX_t D = SVD.singularValues();
     const MatX_t U = SVD.matrixU();
+    reloc_trace::mat("control_U", U);
+    reloc_trace::mat("control_D", D);
 
     for (unsigned int i = 1; i < 4; ++i) {
         const double k = std::sqrt(D(i - 1, 0) / num_correspondences);
@@ -260,6 +284,7 @@
         }
     }
     const Mat33_t CC_inv = svd.matrixV() * S * svd.matrixU().transpose();
+    reloc_trace::mat("CC_inv", CC_inv);
 
     eigen_alloc_vector<Vec4_t> alphas;
     for (unsigned int i = 0; i < num_correspondences; ++i) {
@@ -370,6 +395,9 @@
     Eigen::JacobiSVD<MatX_t> SVD(CM, Eigen::ComputeFullV | Eigen::ComputeFullU);
     const MatX_t& CM_u = SVD.matrixU();
     const MatX_t& CM_vt = SVD.matrixV().transpose();
+    reloc_trace::mat("CM", CM);
+    reloc_trace::mat("CM_U", CM_u);
+    reloc_trace::mat("CM_Vt", CM_vt);
 
     rot = CM_u * CM_vt;
     const double det = rot.determinant();
@@ -414,6 +442,7 @@
 
     Eigen::JacobiSVD<MatX_t> SVD(L_6x3, Eigen::ComputeFullV | Eigen::ComputeFullU);
     const Vec3_t b3 = SVD.solve(Rho);
+    reloc_trace::mat("beta_solution", b3);
 
     if (b3(0) < 0) {
         betas(0) = std::sqrt(-b3(0));
@@ -451,6 +480,7 @@
 
     Eigen::JacobiSVD<MatX_t> SVD(L_6x5, Eigen::ComputeFullV | Eigen::ComputeFullU);
     const Vec5_t b5 = SVD.solve(Rho);
+    reloc_trace::mat("beta_solution", b5);
 
     if (b5(0) < 0) {
         betas(0) = std::sqrt(-b5(0));
@@ -483,6 +513,7 @@
 
     Eigen::JacobiSVD<MatX_t> SVD(L_6x4, Eigen::ComputeFullV | Eigen::ComputeFullU);
     const Vec4_t b4 = SVD.solve(Rho);
+    reloc_trace::mat("beta_solution", b4);
 
     if (b4(0) < 0) {
         betas(0) = std::sqrt(-b4(0));
@@ -571,9 +602,12 @@
 
     for (unsigned int j = 0; j < num_iter; j++) {
         compute_A_and_b_for_gauss_newton(L_6x10, Rho, betas, A, B);
+        reloc_trace::mat("GN_A", A);
+        reloc_trace::mat("GN_B", B);
 
         // Using fastest QR decomposition in Eigen
         betas += A.householderQr().solve(B);
+        reloc_trace::mat("GN_betas", betas);
     }
     return betas;
 }
